bfvmm/include/hve/arch/intel_x64/vmexit/cpuid.h

```diff
@@ -149,53 +128,109 @@ class EXPORT_HVE cpuid_handler
+    /// an emulator handler tells the APIs that you are taking on the
+    /// responsibility of properly handling the hardware, including ensuring
+    /// that the hardware (or fake hardware) is consistent with what the base
+    /// hypervisor provides, including any assumptions it is making. Use wisely.
+    ///
+    /// To handle the VM exit, modify the cr4 register using the following two
+    /// functions as needed:
+    /// - vcpu->rax()
+    /// - vcpu->set_rax()
+    /// - vcpu->rbx()
+    /// - vcpu->set_rbx()
+    /// - vcpu->rcx()
+    /// - vcpu->set_rcx()
+    /// - vcpu->rdx()
+    /// - vcpu->set_rdx()
     ///
     /// @expects
     /// @ensures
     ///
     /// @param leaf the address to emulate
+    /// @param d the handler to call when an exit occurs
     ///
-    void emulate(leaf_t leaf);
+    void add_emulator(leaf_t leaf, const handler_delegate_t &d);
 
-    /// Add Default Handler
+    /// Execute
     ///
-    /// This is called when no registered handlers have been called and
-    /// the internal implementation is needed. Note that this function
-    /// can still return false and let the internal implementation pass
-    /// the instruction through
+    /// Executes the CPUID instruction and populates the vCPU's registers as
+    /// follows:
+    /// - gr1 = vcpu->rax() (leaf input)
+    /// - gr2 = vcpu->rcx() (subleaf input)
+    /// - [rax, rbx, rcx, rdx] = cpuid
+    /// - vcpu->rax() = rax (output)
+    /// - vcpu->rbx() = rbx (output)
+    /// - vcpu->rcx() = rcx (output)
+    /// - vcpu->rdx() = rdx (output)
     ///
-    /// Also note that the handler registered here is a base exit handler
-    /// delegate. The info structure is not passed, and therefor,
-    /// no emulation is provided to this handler.
+    /// Note: This function can be used inside of an emulator to
+    /// access hardware similar to a regular handler. This is useful when you
+    /// want to use an emulator, but still need to access hardware. Just be
+    /// aware that the safety protections that an emulator provides are
+    /// removed.
     ///
     /// @expects
     /// @ensures
     ///
-    /// @param d the handler to call when an exit occurs
+    /// @param vcpu the vcpu object to execute CPUID on
+    ///
+    void execute(gsl::not_null<vcpu *> vcpu);
+
+    /// Enable Whitelisting
+    ///
+    /// By default, if an emulator is not registered, the base implementation
+    /// will handle the VM exit for you automatically. If whitelisting is
+    /// enabled, this behavior is disabled, and the base implementation will
+    /// report the vm exit as unhandled, generating an exception. This is
+    /// mostly useful for guest vCPUs that wish to halt() the vCPU if a
+    /// register is accessed that is not explicitly being emulated.
+    ///
+    /// @expects
+    /// @ensures
     ///
-    void set_default_handler(const ::handler_delegate_t &d);
+    void enable_whitelisting() noexcept
```

Comment:
What is it that we're whitelisting?

Reply:
CPUID. Once you turn this on, you have to produce a whitelist of valid CPUID leaves. If you have not registered an emulator for a leaf, it will be reported as unhandled. If this is not enabled, the implementation will fallback to registered handlers. This is needed to ensure that guest vCPUs explicitly emulate all hardware.

Reply:
So if a leaf is whitelisted, that means that you don't need an emulator for it?